Topics/Tech_Stacks/Writing_Clean_React_Code.md

```diff
@@ -0,0 +1,153 @@
+
+```jsx
+const ParentComponent = () => {
+  const handleClick = () => {
+    // handle button click
+  };
+  return <Button text="Click Me" onClick={handleClick} />;
+};
+```
+
+- **Constants**: Constants should be named using uppercase letters and underscores to separate words. For example, a constant that stores user emails should be named `USER_EMAILS` instead of `user-emails` or `userEmails`.
+
+- **File Names**: File names should be named using PascalCase. For example, a file that contains a component that renders a navbar element should be named `NavbarElement.jsx` instead of `navbarElement.jsx` or `navbar-element.jsx`.
+
+- **CSS Classes and Files**: CSS classes should be named using kebab-case or camelCase. For example, a class that styles a div that acts as a container for a cards should be named `cards-container` or `cardsContainer` instead of `cards_container`. CSS Files follows the same convention as CSS classes.
+
+As to what you should name your classes, a popular convention is to use the BEM (Block, Element, Modifier) naming convention. This convention is useful for naming classes that are used to style components. It is a pretty big topic in its own right, so to learn more you might want to check out the [getbem.com website](https://getbem.com/introduction/).
+
+## Component Structure
+
+When working with React, it is important to follow a consistent component structure. This helps developers quickly identify the purpose of each component and its role in the application. The following are some common principles for structuring React components:
+
+- **Break down components**: Components should be broken down into smaller components whenever possible. Large components are difficult to understand and debug. It is also easier to reuse smaller components, which avoids code repetition. This essentially goes back to the [Single Responsibility Principle](https://en.wikipedia.org/wiki/Single-responsibility_principle) in software engineering, which states that each component should have a single responsibility. For example, a component that renders a user profile page should be broken down into smaller components that render the user's name, profile picture, and bio.
+
+- **Organize component heirarchy into folders**: Components should be organized into folders based on their role in the application. At the highest level inside the components folder, there should be folders for each of the pages in the website. Within each page folder, there should be folders for each of the components that make up the page. The more deeply nested a component is within its page, the deeper it will be in the folder heirarchy. Moreover, test and CSS files should be together with the jsx files. For example, a website that has a home page and a user profile page should have the following folder structure:
```

Comment:
nit: Typo - third word should be hierarchy.

Reply:
My mistake. Fixed!